### ISSUE TITLE
[Bug]: `Audio.setSinkId` fails if run after `navigator.mediaDevices.getUserMedia` on audio inputs

### ISSUE THREAD

**Issue by smallhadroncollider** [electron/electron#40704]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

27.0.4

### What operating system are you using?

macOS

### Operating System Version

macOS Sonoma 14.1 (23B74)

### What arch are you using?

arm64 (including Apple Silicon)

### Last Known Working Electron version

23.3.0

### Expected Behavior

- Run `getUserMedia` on all inputs
- Switch system audio output to BlackHole driver
- Use `setSinkId` to output the BlackHole audio stream to speakers
- System audio should output via the speakers

### Actual Behavior

In all version of Electron after 23 the audio does not output over the speakers. If you switch output from BlackHole to speakers and then _back_ to BlackHole it starts working as expected.

### Testcase Gist URL

https://gist.github.com/smallhadroncollider/1d1d16141f35617ed4a34107786dae38

### Additional Information

- Had same issue on Intel Mac
- Gist requires installing [BlackHole Audio Driver](https://github.com/ExistentialAudio/BlackHole) (`brew install blackhole-2ch`)
- Your speaker name may be different
- Artificial 10 second delay to switch output devices

**Comment by ckerr** [comment]
@smallhadroncollider does this require the BlackHole driver in order to reproduce, or is that intrinsic to the issue?

Would it be possible to create a repro testcase without 3rd party software?

**Comment by smallhadroncollider** [comment]
@ckerr Hello, sorry for delay getting back to you - been a busy week.

I've update the Gist with a few more details: https://gist.github.com/smallhadroncollider/1d1d16141f35617ed4a34107786dae38

I think it _has_ to be something like BlackHole (the [Electron website recommends SoundFlower](https://www.electronjs.org/docs/latest/api/desktop-capturer#caveats)). I had thought we could use headphones to recreate the issue, but it needs to be a device that works as an input _and_ an output, otherwise the issue doesn't occur.

We've tried it against various versions of BlackHole and it doesn't seem to make any difference. Seems to be something has changed in Electron `> 23` that causes the issue.

---

I was so glad to find a test case, that I didn't mention another aspect of it.

The issue seems to be that if we get the BlackHole _input_ `MediaStream` and then try and get it again after that we can no longer sink the audio to the speakers (everything else seems to work fine). In the Fiddle that's done by iterating over all the inputs. We do this in our app to check that they actually are available (on some Windows machines they're listed but don't actually work). My initial thought was to turn this check off on macOS, as we only needed to check it on Windows anyway. This meant that the first time a user recorded system audio, everything worked fine (🥳), however, if afterwards the user started a _second_ session, we'd get the same issue where the audio sink didn't work and they couldn't hear anything through their speakers.
